src/security/context.cpp

```diff
@@ -643,16 +694,197 @@ class PolFinalWafCtx : public PolTaskCtx<PolFinalWafCtx> {
 
   void complete() noexcept {
     ngx_log_debug0(NGX_LOG_DEBUG_HTTP, req_.connection->log, 0,
-                   "completion handler of waf task");
+                   "completion of final waf task");
   }
 
   friend PolTaskCtx;
 };
 
+ngx_int_t Context::do_request_body_filter(ngx_http_request_t &request,
+                                          ngx_chain_t *in, dd::Span &span) {
+  auto st = stage_->load(std::memory_order_acquire);
+  ngx_log_debug3(
+      NGX_LOG_DEBUG_HTTP, request.connection->log, 0,
+      "request body filter %s in chain. Current data: %lu, Stage: %d",
+      in ? "with" : "without", filter_ctx_.out_total, st);
+
+  if (st == stage::AFTER_BEGIN_WAF) {
+    ngx_log_debug1(NGX_LOG_DEBUG_HTTP, request.connection->log, 0,
+                   "first filter call, req refcount=%d", request.main->count);
+    // buffer the request body during reading
+    // https://github.com/nginx/nginx/commit/67d160bf25e02ba6679bb6c3b9cbdfeb29b759de
+    // https://nginx.org/en/docs/dev/development_guide.html#http_request_body_filters
+    // this essentially avoids early req termination if buffers are not read
+    // (position advanced) by the filters. However, nginx doesn't keep calling
+    // the filters in that case. We use it to avoid synchronous calls to the
+    // filter after we collected enough data to call the WAF. Asynchronous calls
+    // are avoided by swapping the handlers before starting the WAF task.
+    request.request_body->filter_need_buffering = true;
+
+    if (in && in->buf->pos ==
+                  request.header_in->pos - (in->buf->last - in->buf->pos)) {
+      // preread call by ngx_http_read_client_request_body.
+      // read and write handlers were not set yet, so don't launch the
+      // waf now. We'll do it on the next call.
+      ngx_log_debug0(NGX_LOG_DEBUG_HTTP, request.connection->log, 0,
+                     "1st preread call, no waf task submission yet");
+      st = transition_to_stage(stage::COLLECTING_ON_REQ_DATA_PREREAD);
+    } else {
+      st = transition_to_stage(stage::COLLECTING_ON_REQ_DATA);
+    }
+  }
+
+  if (st == stage::COLLECTING_ON_REQ_DATA_PREREAD) {
+    // we're guaranteed to be called again synchronously, so we shouldn't
+    // call the WAF at this point
+
+    if (buffer_chain(*request.pool, in, true) != NGX_OK) {
+      return NGX_HTTP_INTERNAL_SERVER_ERROR;
+    }
+
+    st = transition_to_stage(stage::COLLECTING_ON_REQ_DATA);
+  } else if (st == stage::COLLECTING_ON_REQ_DATA) {
+    // check if we have enough data to run the WAF
+    size_t new_size = filter_ctx_.out_total;
+    bool is_last = filter_ctx_.found_last;
+    for (auto *cl = in; cl; cl = cl->next) {
+      new_size += cl->buf->last - cl->buf->pos;
+      is_last = is_last || cl->buf->last_buf;
+    }
+
+    bool run_waf = new_size >= kMaxFilterData || is_last;
+    if (run_waf) {
+      // do not consume the buffer so that this filter is not called again
```

Comment:
From what I can see this can only potentially be a problem if we get data while suspended (should not happen, and if it does already causes synchronisation issues and possible corruption). Otherwise the data is either copied and consumed or passed downstream immediately (where it will likely get to `ngx_http_request_body_save_filter` and be consumed there).

I guess I could add a condition to make it consume the buffer if the `recycled` flag is true, but if I need to get the data-on-suspended-state anomalous situation to work correctly I also need to add a separate chain to avoid synchronization issues.